### DISCUSSION TITLE
Dictionaries: Design and Validation

### DISCUSSION THREAD

**Discussion by petermr** [petermr/petermr#40] [Ideas]
# Overview

This channel will collect ideas, specifications, implementations, editors and validators for dictionaries. Specifically:

## Principles

What are the dictionaries for? Who will create them? Who will use them?

## Specification

There must be a formal, precise, implementable specification which should be easy to follow.

## Tools 

All specifications must be validatable by running programs or schemas (Schematron, not XSD)

**Comment by petermr** [comment]
# Principles

## Anyone can create a dictionary

## Dictionaries should use a **simple** subset of XML

## Dictionaries must be allowed to evolve dynamically

## Dictionaries should use standards if possible

## Dictionaries will be closely linked to Wikimedia, especially Wikidata

## Dictionaries must welcome global diversity, especially in (human) languages 

## Dictionaries should be part of a growing informal ecosystem

## Ease of entry is more important than formal correctness

**Comment by petermr** [comment]
# Specification

## Dictionaries use a simple subset of XML

## The primary validator is a Python program

## There are a small subset of mandatory elements:
### `dictionary`

### `entry`

## There are a (growing) number

**Comment by EmanuelFaria** [comment]
# Dictionary Standard
================
## Ideas
- Can we automatically use stemming and lemmatization to ensure all inflected forms of a term are captures/detected? This would be especially useful in dictionaries involving activities. Terms like "x promotion/promoting agent"... "promotes/promoted x"... 
 
## GENERAL RULES

- Dictionaries must be *well-formed* XML but are not schema-valid.
	- Content within xml tags that include quotes `(eg. <entry term=“…”>)` cannot have odd numbers of quotation marks (” ”)
- The dictionary has no namespace.
- Common XML languages (HTML, SVG) can be used without namespaces (name collisions are avoided by design


----

# elements


## term


### attribute of `entry`

- Required: YES
- terms are generally EN and generally use `[A-Za-z0-9_-]+`
- Other punctuation should be avoided (`',()` etc) but may be required (e.g. for  chemicals)
- Characters above 128 (diacritics, symbols, ideographs, etc.) must be encoded as Unicode
- Greek characters should be replaced by `alpha`, etc but *may* be used in synonyms
- URL-encoding (e.g. %20) or XML entities (`&amp;`) should not be used
- terms may have synonyms for searching
- terms should be trimmed and multiword terms should include single-spaces.
- terms are generally flattened to lowercase but some must retain  uppercase (e.g.  UN, SDGs)

## desc


### element

- child of `meta` 
- Required: OPTIONAL

### element

- child of `entry`
- Required: OPTIONAL
- Newlines in descriptions can logically be replaced by space (e.g. for tools with flowing text).
- Descriptions should not include line breaks, paragraphs, lists, etc.)
- descriptions may contain well-formed HTML markup but cannot rely on it being honoured.
- **content format**
	- must have at least one entry
		- (Is there a max limit (as spreadsheets have)?

			- If, so.. can we split automatically in pt1, pt2 etc? Can those be “run” concurrently with results showing as if they were one dictionary?)


		- must not include quotes within the xml formatting


----

# meta


## encoding

- child of `meta` 
- Required: REQUIRED
- <?xml version="1.0" encoding="UTF-8"?>

## title

- child of `meta` 
- Required: REQUIRED
- **title** must be match **dictionary’s file name**

## version

- child of `meta` 
- Required: OPTIONAL
- Format: numbers and decimals (0.0.0)

## modificationDate

- child of `meta` 
- Required: OPTIONAL
- Format: date?

## modifiedBy

- child of `meta` 
- Required: OPTIONAL
- Format: text (could be a name of a person or program

## modificationDetails

- child of `meta` 
- Required: OPTIONAL
- Format: text

## authors

- child of `meta` 
- Required: OPTIONAL
- Format: text

## contributors

- child of `meta` 
- Required: OPTIONAL
- Format: text

## datasource(s)

- child of `meta` 
- Required: OPTIONAL
	- I think of this as dictionaries “translated” from existing sources such as Ontobee Ontologies, proprietary databases, webscraped, Dr. Duke, MCID
- Format: text, may be URLs
- <datasource>”…"</datasource>
- eg [http://www.nipgr.ac.in/Essoildb](http://www.nipgr.ac.in/Essoildb), FDA, Chebli

## license

- child of `meta` 
- Required: OPTIONAL
- Format: text, numbers, symbols ™ ® ©  (Possibly not roman text)
- **examples**
	- [https://creativecommons.org/licenses/by-nc/4.0/?ref=chooser-v1](https://creativecommons.org/licenses/by-nc/4.0/?ref=chooser-v1)
	- [https://github.com/petermr/ami3/blob/master/LICENSE](https://github.com/petermr/ami3/blob/master/LICENSE)


----

# entry


## wikidata_ID

- child of `entry` 
- Required: OPTIONAL
- Format: number

## wikipedia_URL

- child of `entry` 
- Required: OPTIONAL
- Format: URL

## other data or ID sources?

- child of `entry` 
- Required: OPTIONAL
- Format: letters, numbers, possibly dashes, periods or underscores, URL
- Peter, while I was manually creating dictionaries, I sometimes used data from sources such as those below. I didn't include them in the final dictionaries, but should we allow for handling these extra "columns" of data in dictionaries — even if just to strip it out, or move it into a catch-all field?
- examples
	- chebi
	- FDA
	- drDuke
	- ICD‑11
